src/frequenz/sdk/actor/power_distributing/power_distributing.py

```diff
@@ -558,24 +558,100 @@ def _get_components_data(self, batteries: Set[int]) -> List[InvBatPair]:
 
             inverter_id: int = self._bat_inv_map[battery_id]
 
-            battery_data: Optional[BatteryData] = self._battery_receivers[
-                battery_id
-            ].peek()
-
-            if battery_data is None:
-                _logger.warning("None returned from battery receiver %d.", battery_id)
+            data = self._get_battery_inverter_data(battery_id, inverter_id)
+            if data is None:
+                _logger.warning(
+                    "Skipping battery %d because its message isn't correct.",
+                    battery_id,
+                )
                 continue
 
-            inverter_data: Optional[InverterData] = self._inverter_receivers[
-                inverter_id
-            ].peek()
+            pairs_data.append(data)
+        return pairs_data
 
-            if inverter_data is None:
-                _logger.warning("None returned from inverter receiver %d.", inverter_id)
-                continue
+    def _get_battery_inverter_data(
+        self, battery_id: int, inverter_id: int
+    ) -> Optional[InvBatPair]:
+        """Get battery and inverter data if they are correct.
 
-            pairs_data.append(InvBatPair(battery_data, inverter_data))
-        return pairs_data
+        Each float data from the microgrid can be "NaN".
+        We can't do math operations on "NaN".
+        So check all the metrics and:
+        * if power bounds are NaN, then try to replace it with the corresponding
+        power bounds from the adjacent component. If metric in the adjacent component is
+        also NaN, then return None.
+        * if other metrics are NaN then return None. We can't assume anything for other
+        metrics.
```

Comment:
Not important as this is private and it won't get rendered, but just a reminder for the cases where it will:

```suggestion
        We can't do math operations on "NaN".        
        So check all the metrics and:
        * if power bounds are NaN, then try to replace it with the corresponding
          power bounds from the adjacent component. If metric in the adjacent component is
          also NaN, then return None.
        * if other metrics are NaN then return None. We can't assume anything for other
          metrics.
```

Reply:
Done